src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,163 @@
+
+/-- The Bernoulli polynomials are defined in terms of the negative Bernoulli numbers. -/
+def bernoulli_poly (n : ℕ) : polynomial ℚ :=
+  ∑ i in range (n + 1), polynomial.monomial (n - i) ((bernoulli i) * (choose n i))
+
+lemma bernoulli_poly_def (n : ℕ) : bernoulli_poly n =
+  ∑ i in range (n + 1), polynomial.monomial i ((bernoulli (n - i)) * (choose n i)) :=
+begin
+  rw [←sum_range_reflect, add_succ_sub_one, add_zero, bernoulli_poly],
+  apply sum_congr, {refl,},
+  rintros x hx,
+  rw mem_range_succ_iff at hx, rw [choose_symm hx, nat.sub_sub_self hx],
+end
+
+namespace bernoulli_poly
+
+/-
+### examples
+-/
+
+section examples
+
+@[simp] lemma bernoulli_poly_zero : bernoulli_poly 0 = 1 :=
+by { rw bernoulli_poly, simp, }
+
+@[simp] lemma bernoulli_poly_zero' (n : ℕ) : (bernoulli_poly n).eval 0 = bernoulli n :=
+begin
+  rw [bernoulli_poly, polynomial.eval_finset_sum], simp only [polynomial.eval_monomial],
+  rw sum_range_succ,
+  simp only [add_right_eq_self, mul_one, cast_one, nat.sub_self, choose_self, pow_zero],
+  apply sum_eq_zero, rintros x hx,
+  rw [zero_pow', mul_zero], apply mem_range_sub_ne_zero hx,
+end
+
+end examples
+
+@[simp] theorem sum_bernoulli_poly (n : ℕ) :
+  ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k =
+    polynomial.monomial n (n + 1 : ℚ) :=
+begin
+  cases n,
+  { simp only [bernoulli_poly_zero, cast_one, choose_succ_self_right, one_smul,
+      polynomial.monomial_zero_left, cast_zero, sum_singleton, zero_add, ring_hom.map_one,
+      range_one], },
+  simp_rw [bernoulli_poly_def, finset.smul_sum, finset.range_eq_Ico, ←finset.sum_Ico_Ico_comm,
+    finset.sum_Ico_eq_sum_range],
+  simp only [cast_succ, nat.add_sub_cancel_left, nat.sub_zero, zero_add, linear_map.map_add],
+  simp_rw [polynomial.smul_monomial, mul_comm (bernoulli _) _, smul_eq_mul, ←mul_assoc],
+  conv_lhs { apply_congr, skip, conv
+    { apply_congr, skip,
+      rw [nat.prime.choose_mul ((nat.le_sub_left_iff_add_le (mem_range_le H)).1 (mem_range_le H_1))
+        (le.intro rfl), add_comm x x_1, nat.add_sub_cancel, mul_assoc, mul_comm, ←smul_eq_mul,
+        ←polynomial.smul_monomial], },
+    rw [←sum_smul], },
+  rw sum_range_succ,
+  simp only [add_right_eq_self, cast_succ, mul_one, cast_one, cast_add, nat.add_sub_cancel_left,
+    choose_succ_self_right, one_smul, bernoulli_zero, sum_singleton, zero_add,
+    linear_map.map_add, range_one],
+  have f : ∀ x ∈ range n.succ, 2 ≤ n.succ + 1 - x,
+  { rintros x H,
+    rw [succ_sub (ge_iff_le.1 (mem_range_le H)), succ_le_succ_iff,
+      succ_sub (ge_iff_le.1 (mem_range_succ_iff.1 H)), succ_le_succ_iff], norm_num, },
+  conv_lhs { apply_congr, skip, rw [sum_bernoulli _ (f x H), zero_smul], },
+  simp,
+end
+
+open power_series
+variables {A : Type*} [integral_domain A] [algebra ℚ A] [char_zero A]
+
+theorem exp_bernoulli_poly (t : A) :
+  power_series.mk (λ n, polynomial.aeval t ((1 / nat.factorial n : ℚ) • bernoulli_poly n))
+    * (exp A - 1) = X * rescale t (exp A) :=
```

Comment:
I don't necessarily want to hold things up, but this proof is at least the length that internal comments would be very helpful.

Reply:
Sure, I'll add those.

Reply:
Done!